pwiz_tools/Skyline/Model/Results/SpectrumFilter.cs

```diff
@@ -1,1149 +1,1298 @@
+            if (instrumentInfo != null)
+            {
+                _isWatersFile = instrumentInfo.IsWatersFile;
+                _configInfoList = instrumentInfo.ConfigInfoList;
+            }
+            IsFirstPass = firstPass;
+
+            var comparer = PrecursorTextId.PrecursorTextIdComparerInstance;
+            var dictPrecursorMzToFilter = new SortedDictionary<PrecursorTextId, SpectrumFilterPair>(comparer);
+
+            // If we're using bare measured ion mobility values from spectral libraries, go get those now
+            var libraryIonMobilityInfo = document.Settings.PeptideSettings.Prediction.UseLibraryIonMobilityValues
+                ? document.Settings.GetIonMobilities(msDataFileUri)
+                : null;
+            var ionMobilityMax = maxObservedIonMobilityValue ?? 0;
+
+            // TIC and Base peak are meaningless with FAIMS, where we can't know the actual overall ion counts -also can't reliably share times with any ion mobility scheme
+            if (instrumentInfo != null && instrumentInfo.IonMobilityUnits != eIonMobilityUnits.none)
+            {
+                foreach (var pair in document.MoleculePrecursorPairs)
+                {
+                    double windowIM;
+                    var ionMobility = document.Settings.GetIonMobility(
+                        pair.NodePep, pair.NodeGroup, null, libraryIonMobilityInfo, _ionMobilityFunctionsProvider, ionMobilityMax, out windowIM);
+                    _isIonMobilityFiltered = ionMobility.HasIonMobilityValue;
+                    if (_isIonMobilityFiltered)
+                    {
+                        break;
+                    }
+                }
+            }
+
+            if (EnabledMs || EnabledMsMs)
+            {
+                if (EnabledMs)
+                {
+                    _isHighAccMsFilter = !Equals(_fullScan.PrecursorMassAnalyzer,
+                        FullScanMassAnalyzerType.qit);
+
+                    /*
+                     Leaving this here in case we ever decide to fall back to our own BPC+TIC extraction in cases where data
+                     file doesn't have them ready to go, as in mzXML
+                    if (!firstPass && _isIonMobilityFiltered)
+                    {
+                        var key = TIC_KEY;
+                        dictPrecursorMzToFilter.Add(key, new SpectrumFilterPair(key, PeptideDocNode.UNKNOWN_COLOR, dictPrecursorMzToFilter.Count,
+                            _instrument.MinTime, _instrument.MaxTime, _isHighAccMsFilter, _isHighAccProductFilter));
+                        key = BPC_KEY;
+                        dictPrecursorMzToFilter.Add(key, new SpectrumFilterPair(key, PeptideDocNode.UNKNOWN_COLOR, dictPrecursorMzToFilter.Count,
+                            _instrument.MinTime, _instrument.MaxTime, _isHighAccMsFilter, _isHighAccProductFilter));
+                    }
+                    //*/
+                }
+                if (EnabledMsMs)
+                
```

Comment:
We usually combine mass errors by doing an average weighted by the observed intensity (so, you would keep a running total of mass error times intensity, and then divide by total instensity.

Reply:
So more like this?
`            if (is_diaPASEF)
            {
                // We may have worked through several isolation windows all at the same retention time - sum their intensities.
                foreach (var kvp in diaPasefFilteredSpectra)
                {
                    var totalIntensity = 0.0f;
                    var filteredSrmSpectrum = kvp.Value.First();
                    foreach (var extractedSpectrum in kvp.Value.Skip(1))
                    {
                        for (var i = 0; i < filteredSrmSpectrum.Intensities.Length; i++)
                        {
                            var extractedSpectrumIntensity = extractedSpectrum.Intensities[i];
                            filteredSrmSpectrum.Intensities[i] += extractedSpectrumIntensity;
                            filteredSrmSpectrum.MassErrors[i] += extractedSpectrumIntensity * extractedSpectrum.MassErrors[i];
                            totalIntensity += extractedSpectrumIntensity;
                        }
                    }
                    if (totalIntensity != 0)
                    {
                        for (var i = 0; i < filteredSrmSpectrum.Intensities.Length; i++)
                        {
                            filteredSrmSpectrum.MassErrors[i] /= totalIntensity; // Average the mass errors
                        }
                    }
                    yield return filteredSrmSpectrum;
                }
            }
`

Reply:
wow, that formatted horribly, sorry
`            if (is_diaPASEF)

            {

                // We may have worked through several isolation windows all at the same retention time - sum their intensities.

                foreach (var kvp in diaPasefFilteredSpectra)

                {

                    var totalIntensity = 0.0f;

                    var filteredSrmSpectrum = kvp.Value.First();

                    foreach (var extractedSpectrum in kvp.Value.Skip(1))

                    {

                        for (var i = 0; i < filteredSrmSpectrum.Intensities.Length; i++)

                        {

                            var extractedSpectrumIntensity = extractedSpectrum.Intensities[i];

                            filteredSrmSpectrum.Intensities[i] += extractedSpectrumIntensity;

                            filteredSrmSpectrum.MassErrors[i] += extractedSpectrumIntensity * extractedSpectrum.MassErrors[i];

                            totalIntensity += extractedSpectrumIntensity;

                        }

                    }

                    if (totalIntensity != 0)

                    {

                        for (var i = 0; i < filteredSrmSpectrum.Intensities.Length; i++)

                        {

                            filteredSrmSpectrum.MassErrors[i] /= totalIntensity; // Average the mass errors

                        }

                    }

                    yield return filteredSrmSpectrum;

                }

            }

`